### PR TITLE
Add QNN (Qualcomm HTP) backend to the ExecuTorch exporter

### PR DESCRIPTION
**Pull request by psiddh**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46989)

Shift-left of QNN export from optimum-executorch into transformers: exporting an HF model with ExecutorchConfig(backend="qnn") now produces a QNN-delegated .pte directly.

New src/transformers/exporters/qnn_utils.py with three hooks, wired into the exporter's backend dispatch:
- prepare_for_qnn: pre-export prep — static-KV-cache wrap for decoder-only LMs (decode-capable graph; avoids the constant-folded-mask "No graph inputs present" QNN serialization failure) + HTP compiler spec (fp16, or int for the quantized path).
- quantize_for_qnn (_POST_EXPORT_TRANSFORM): PT2E 8a8w/16a4w quantization AFTER torch.export, which is where QNN quantizes (QnnQuantizer annotates the exported graph, then convert_pt2e).
- _lower_for_qnn (_BACKEND_LOWER): backend-specific lowering that runs QNN's pre-export + edge pass pipeline inside a QnnManagerContext. The exporter's generic to_edge_transform_and_lower can't lower QNN (hits KeyError aten.alias_copy.default / rank-3 FullyConnected / "No QnnManager active") because those passes never run.

Exporter plumbing:
- Add a _POST_EXPORT_TRANSFORM hook (runs on the ExportedProgram between export and lowering) and a _BACKEND_LOWER table (backend supplies its own lowering; default stays the generic path).
- ExecutorchConfig gains soc_model / quantize / static_cache / max_cache_len / calibration_inputs (all QNN-only; ignored by xnnpack/cuda).
- integrations/executorch.py: TorchExportableModuleForDecoderOnlyLM.forward now threads cache_position into the static-cache module (it was dropped, breaking the decode graph).

Validated on Qwen2.5-0.5B -> fully QnnBackend-delegated .pte (FP16, SM8750 and SM8550).

# What does this PR do?

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [ ] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/exporters/configs.py [hunk 1]
```diff
@@ -163,8 +163,28 @@ class ExecutorchConfig(DynamoConfig):
...
     """
 
     export_format: ExportFormat = ExportFormat.EXECUTORCH
 
     backend: str = "xnnpack"
+    soc_model: str = "SM8750"
+    quantize: str | None = None
+    static_cache: bool = True
+    max_cache_len: int = 128
+    calibration_inputs: list[Any] | None = None
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/configs.py:190]: i don't think we wanna support quantization as part of the hf exporters api, this is very "specific" while the approach here is to be as general as possible

File: src/transformers/exporters/exporter_executorch.py [hunk 2]
```diff
@@ -198,6 +208,39 @@ def prepare_for_cuda(model: PreTrainedModel, sample_inputs: dict[str, Any]):
...
+# ── Post-export transforms ────────────────────────────────────────────────────
+# Run on the ExportedProgram between torch.export.export() and lowering. XNNPACK quantizes
+# TorchAO-eager *before* export (nothing here); QNN quantizes with PT2E *after* export
+# (QnnQuantizer annotates the exported graph for HTP), so it registers a hook below.
+# Signature: transform(exported_program, sample_inputs, config) -> ExportedProgram.
+
+
+_POST_EXPORT_TRANSFORM = {
+    "qnn": quantize_for_qnn,
+}
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_executorch.py:224]: should be done by another api, not exporters tbh

File: src/transformers/exporters/configs.py (hunk 1 above)
Comment (psiddh) [reply src/transformers/exporters/configs.py:190]: Agree! Would a generic backend_config: dict work instead of QNN-specific fields, or should backend config also live outside the exporter?

File: src/transformers/exporters/exporter_executorch.py (hunk 2 above)
Comment (psiddh) [reply src/transformers/exporters/exporter_executorch.py:224]: Got it, quantization shouldn't be part of the exporter. For backends like Qualcomm , there's a post-export quantization step (prepare → calibrate → convert) that needs to happen between torch.export and lowering. Where do you think that should live? A separate API in transformers, optimum-executorch, or executorch itself?

File: src/transformers/exporters/configs.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/configs.py:190]: honestly as long as it can be a generic recipe covering all models with very little edge case handling, i'm okay with it but in this case it doesn't look like it, i mean it can but we will need some separation of concerns. for example in the exporters, the contract is the user has to know how inference is done with this model, either by being familiar with this model or checking the inference snippet, because they need to know how to create the right inputs for the model, and those inputs are the only thing needed to export it. this can be automated of course like with a model helper model.get_inputs / get_inputs_for_generation but it's not yet decided.

with quantization, if we wanna be the ones creating and handling the dataset it becomes heavy as well. but if we consider that a quantization/calibration dataset is prepared and passed by the user and is just a list of input dicts, that the model can inference with each sample. it becomes simpler. you can take the calibration set, run the model on it, do the same input/output capturing and model decomposition trick and get a calibration dataset per single graph. then each graph can be calibrated using their inferred calibration. this is what i mean by generic.

also it seems that here we are mixing static cache and its specificity with quantization, but the cache should be handled by in a more general way, and the idea is that if you want the model to be exported with static cache, you're definitely gonna/should use export_for_generation which we can make accepts a generation config (like model.generate) that specifies how the generation loop happens (in this case with static cache for example), it should be as simple as passing a generation config to exproter.export_for_generation(model, inputs, export_config, generation_config) wdyt ?

File: src/transformers/exporters/exporter_executorch.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/exporter_executorch.py:224]: i think it depends on how generic can we make it (covers as many models as possible without the incremental per-model additions) and how good are we at establishing a good contract between us and the user. if preparation can be independent from export and export can be independent from calibration, then there is a chance we can put it here but if each ET backend requires a different api it becomes tricky, ofc one way would be to build on top of the ET exporter a QNNExporter which does what it should in the order it wants. but otherwise optimum-executorch can be the fallback since we don't reinforce the constraints there (here almost every feature should be tested fully on every model in our testing suite).
